EIPS/eip-7591.md

```diff
@@ -0,0 +1,114 @@
+---
+eip: 7591
+title: BLS signed transactions
+description: Introduces a new transaction type signed with BLS signatures
+author: Marius van der Wijden (@MariusVanDerWijden)
+discussions-to: https://ethereum-magicians.org/t/eip-7591-bls-signed-transactions/19911
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-01-10
+---
+
+## Abstract
+
+This EIP introduces a new [EIP-2718](./eip-2718.md) transaction type that is signed with BLS signatures.
+
+## Motivation
+
+The BLS signature scheme allows for easy aggregation and verification of aggregated signatures.
+If a substantial number of transactions on mainnet were BLS signed transactions, we can aggregate signatures in a block and batch-verify them.
+This will reduce growth of the chain history.
+
+
+## Specification
+
+### Transaction Type
+
+The transaction type will have the following format: 
+
+```
+[chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, to, value, data, access_list, sender, signature]
+```
+
+with `sender` being the BLS public key of an account with address `address = [0:20](keccak256(sender))`.
+
+The signature value `signature` is calculated by constructing a BLS signature over the following digest:
+
+`tx_hash = keccak256(BLOB_TX_TYPE || rlp([chain_id, nonce, max_priority_fee_per_gas, max_fee_per_gas, gas_limit, to, value, data, access_list, sender]))`.
+
+### Header changes
+
+The block header will be amended with the `aggregated_sig` field, containing an aggregated signature of all BLS transactions in the block.
+
+The resulting RLP encoding of the header is therefore:
+
+```
+rlp([
+    parent_hash,
+    0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347, # ommers hash
+    coinbase,
+    state_root,
+    txs_root,
+    receipts_root,
+    logs_bloom,
+    0, # difficulty
+    number,
+    gas_limit,
+    gas_used,
+    timestamp,
+    extradata,
+    prev_randao,
+    0x0000000000000000, # nonce
+    base_fee_per_gas,
+    withdrawals_root,
+    blob_gas_used,
+    excess_blob_gas,
+    aggregated_sig,
+])
+```
+
+### Block changes
+
+The block building algorithm needs to be changed in order to built the aggregated signature of all BLS signed transactions in the block. 
+All transactions in the block will be added without the signature field set. 
+
+Blocks with transactions containing the `signature` field MUST be rejected.
+
+On block verification, the `verifyAggregate` algorithm is used as follows:
+
+```
+valid = verifyAggregate(sender_1, ... sender_n, tx_hash_1, ... tx_hash_n, aggregated_sig)
```

Comment:
Note that contrary to "batch_verify" we cannot avoid signature forgeries at the cryptographic level with aggregate_verify:
- https://ethresear.ch/t/fast-verification-of-multiple-bls-signatures/5407
- https://ethresear.ch/t/security-of-bls-batch-verification/10748
- https://github.com/status-im/nim-blscurve/blob/d091a57/tests/t_batch_verifier.nim#L146-L156

So either:
- the protocol ensures that the sender_1 pubkey is correct (I think it is)
- or the verification generates "randomness", similar to fiat-shamir construct, by hashing all transactions for example. And then the pubkey and aggregate sig are adjusted to be multiplied by the 64 most significant bits of repeated hash of that randomness, see: https://github.com/mratsim/constantine/blob/f3e08ed/constantine/signatures/bls_signatures.nim#L421-L430

Reply:
[]()